### ISSUE TITLE
The dollar ($) prefixed field '$pullAll' in '$pullAll' is not valid for storage

### ISSUE THREAD

**Issue by smanji** [go-mgo/mgo#87]
I'm using MongoDB 2.6.6 and using the mgo/txn package to do an update on one collection and an insert in another.

Executing this gives the following error: The dollar ($) prefixed field '$pullAll' in '$pullAll' is not valid for storage

Per the following thread, "$ is not allowed at the beginning of a field name, since in this position it is reserved to indicate query operators like $or, $addToSet, $, etc, and allowing it as part of field names would make the query language ambiguous":
https://groups.google.com/forum/#!topic/mongodb-user/uViuJsl4jP8

Can this package be updated to support Mongo 2.6?

Thanks!

**Comment by smanji** [comment]
I think I fixed this issue with the following pull request: https://github.com/go-mgo/mgo/pull/88

Please review and if you agree, merge it into the original and we can close this issue.

Thanks!
Sadiq.

**Comment by niemeyer** [comment]
The txn package works in MongoDB 2.4 and 2.6, and probably earlier ones. The error you describe doesn't make sense to me, as it reports an attempt to store a $ field, which $pullAll should of course not be as this is an update operator.  The pull request also does not make sense, as it is turning such an update operator into a normal field, thus breaking the package (there is a test suite you can run to test your changes).

Can you please provide a tiny and self-contained test case demonstrating the failure you observed?

**Comment by smanji** [comment]
Hi Gustavo,

Thanks for your response. I'm running mongodb 2.6.6 on Mac OSX.

Here's the tiniest test case I could create to reproduce my issue, it's the second transaction that gives the error -- hopefully you can point out what I'm doing wrong.

Incidentally, my fix in pull request #88 resolves this issue for me... sorry for not testing it with your test suite though!

```Go
package main

import (
	"fmt"
	"gopkg.in/mgo.v2"
	"gopkg.in/mgo.v2/bson"
	"gopkg.in/mgo.v2/txn"
	"time"
)

type ParentStruct struct {
	Id          bson.ObjectId `json:"id,omitempty" bson:"_id"`
	UpdatedDate time.Time     `json:"updated_date" bson:"updated_date,omitempty"`
}

type ChildStruct struct {
	Id       bson.ObjectId `json:"id,omitempty" bson:"_id"`
	ParentId bson.ObjectId `json:"parent_id" bson:"parent_id"`
}

func main() {
	mgoSession, _ := mgo.Dial("localhost")
	db := mgoSession.DB("test")
	txnCollection := db.C("transactionsCollection")
	parentCollection := "parentCollection"
	childCollection := "childCollection"

	parentId := bson.NewObjectId()
	parentObj := ParentStruct{Id: parentId, UpdatedDate: time.Now().UTC()}

	childObj1 := ChildStruct{Id: bson.NewObjectId(), ParentId: parentId}
	childObj2 := ChildStruct{Id: bson.NewObjectId(), ParentId: parentId}

	runner_txn1 := txn.NewRunner(txnCollection)
	ops_txn1 := []txn.Op{{
		C:      parentCollection,
		Id:     parentObj.Id,
		Insert: parentObj,
	}, {
		C:      childCollection,
		Id:     childObj1.Id,
		Insert: childObj1,
	}}
	id_txn1 := bson.NewObjectId() // Optional
	err_txn1 := runner_txn1.Run(ops_txn1, id_txn1, nil)

	if err_txn1 != nil {
		fmt.Println("Error inserting parent and child: %v", err_txn1.Error())
		return
	}

	parentObj.UpdatedDate = time.Now().UTC()
	runner_txn2 := txn.NewRunner(txnCollection)
	ops_txn2 := []txn.Op{{
		C:      parentCollection,
		Id:     parentObj.Id,
		Update: parentObj,
	}, {
		C:      childCollection,
		Id:     childObj2.Id,
		Insert: childObj2,
	}}
	id_txn2 := bson.NewObjectId() // Optional
	err_txn2 := runner_txn2.Run(ops_txn2, id_txn2, nil)

	if err_txn2 != nil {
		fmt.Println("Error inserting child and updating parent: %v", err_txn2.Error())
		return
	}

	fmt.Println("Successfully processed both transactions!")
}
```

**Comment by niemeyer** [comment]
The documents you provide in the Update field must be $ operations, not the plain objects you inserted.

I'm completely sure your suggested change in #88 is breaking the transaction package. Please run tests after your change and you'll understand.
